---
#date: "This analysis was finished on `r format(Sys.time(), '%Y-%m-%d %H:%M:%S')`"
params:
  args1: "sample_folder"
  args2: "sample_name"
output:
  html_document:
    theme: spacelab
    font-family: Bookman, sans-serif
---

---
title: <b>Alignment summary by STARsolo for the sample `r paste(params$args2)`</b>
author:
  - CEITEC Bioinformatics Core Facility
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
```

```{r load-packages, include= FALSE}
library(data.table)
library(ggplot2)
library(scales)
library(flextable)

knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
set_flextable_defaults(font.family = "Arial", font.size = 12, padding.top = 3, padding.bottom = 3)
```


```{r report formatting, echo = FALSE}

summary <- fread(paste0("../../",params$args1))
summary <- summary[-8,]
summary$V2 <-as.numeric(summary$V2)
num_reads <- summary[1,]
num_reads[,V2 := comma(V2)]
sequencing <- summary[c(2:5,11),]
sequencing[,V2 := paste(V2*100,"%")]
sequencing_summ <- rbind(num_reads, sequencing)
setnames(sequencing_summ, c("V1","V2"), c("Summary","Results"))
mapping_1 <- summary[6:8,]
mapping_1[, V2 := paste(V2*100, "%")]
mapping_2 <- summary[10,]
mapping_2[, V2 := comma(V2)]
mapping_summ <- rbind(mapping_1, mapping_2)
setnames(mapping_summ, c("V1","V2"), c("Summary","Results"))
cell_1 <- summary[11,]
cell_1[,V2 := paste(V2*100, "%")]
cell_2 <- summary[12:19,]
cell_2[,V2:=comma(V2)]
cell_summ <- rbind(cell_1, cell_2)
setnames(cell_summ, c("V1","V2"), c("Summary","Results"))
```

# <u>Sequencing summary</u>

```{r, echo = FALSE}
knitr::kable(sequencing_summ)
```


# <u>Mapping summary</u>

```{r, echo = FALSE}
knitr::kable(mapping_summ)
```

# <u>Cells Summary</u>

```{r, echo = FALSE}
knitr::kable(cell_summ)
```

```{r, results='asis', out.width="80%", echo = FALSE, dev = "svg"}
  cluster_names <- paste0("cluster ", 1:12)
image_paths_umap <- paste0("../../mapped/", params$args2, "/Plots/", params$args2, "_", 1:12, "_UMAP.png")
dt_umap <- data.table(cluster = cluster_names, path = image_paths_umap)

image_paths_tsne <- paste0("../../mapped/", params$args2, "/Plots/", params$args2, "_", 1:12, "_TSNE.png")
dt_tsne <- data.table(cluster = cluster_names, path = image_paths_tsne)
```

### {-}

# <u>Cluster Summary: UMAP</u>
```{r umap-plot,echo=FALSE, out.width="80%", results="asis", dev="svg"}
umap_plot <- function(dt = dt) {
    output<-c(cat("#### Results for 12 UMAP c;usters {.tabset} \n"))
    for (i in 1:length(dt$path)){
      output <- c(output,cat("##### ", dt$cluster[i],"\n"))
      output<-c(output,cat("<div align='center'><figure style='width:80%'><img src=",dt$path[i]," /><figcaption> </figcaption></figure></div>"))
    }
  return(output)
}
umap <- umap_plot(dt_umap)
```

### {-}

# <u>Cluster Summary: t-SNE</u>
```{r tsne-plot,echo=FALSE, out.width="80%", results="asis", dev="svg"}
tsne_plot <- function(dt = dt) {
    output<-c(cat("#### Results for 12 t-SNE clusters {.tabset} \n"))
    for (i in 1:length(dt$path)){
      output <- c(output,cat("##### ", dt$cluster[i],"\n"))
      output<-c(output,cat("<div align='center'><figure style='width:80%'><img src=",dt$path[i]," /><figcaption> </figcaption></figure></div>"))
    }
  return(output)
}
tsne <- tsne_plot(dt_tsne)
```